### PR TITLE
Significant API changes

### PR DESCRIPTION
**Pull request by alxdavids**
The first phase in a large number of changes to the protocol API:

- Define explicit prime-order group API (#81)
- Use concrete data structures for all types
- Provide Serialization/Deserialization mechanisms for groups (#77)
- Ensure that group elements are not exposed by API
- Removes the Setup algorithm (#88, #82)
- Unifies the DLEQ proof algorithms (#85)

Addresses: 

cc @armfazh as initial reporter.

### CODE REVIEW

**Comment by alxdavids** [comment]
We should merge #92 into this PR before we eventually merge this into master.

File: draft-irtf-cfrg-voprf.md [hunk 1]
```diff
@@ -967,369 +814,254 @@ instantiation. We refer to the additive blinding case as fixed-base
...
+    6. Z = di * Zi + Z
+ 6. Output (GG.Serialize(M), GG.Serialize(Z))
+~~~
+
+## Hash function instantiations
+
+In the above algorithmms, we define three new functions `H_3`, `H_4` and
+`H_5`. Note that `H_3` and `H_4` are hash function that maps arbitrary
+length byte outputs to a fixed size output. The function `H_5` should
+output a uniform scalar value in `GF(GG.Order())`. Both functions are
```
Comment (chris-wood) [inline draft-irtf-cfrg-voprf.md:1063]: Can we borrow language from h2c for this? Namely, "should induce an output distribution in GF(GG.Order()) is statistically close to uniformly random."

File: draft-irtf-cfrg-voprf.md [hunk 2]
```diff
@@ -455,36 +488,23 @@ with a prime-order group instantiation are removed. In the case of
...
-
-- We use the notation x <-$ Q to denote sampling x from the uniform
-  distribution over the set Q.
-- We use x <- {0,1}^u to denote sampling x uniformly from the set of
-  binary strings of length u. We may interpret x afterwards as a byte
-  array.
-- We say that x is a binary string of arbitrary-length (or alternatively
-  sampled from {0,1}^*) if there is no fixed-size requirement on x.
-- For two byte arrays x and y, write x .. y to denote their concatenation.
+Secure instances of such groups can also be constructed by choosing `GG`
```
[the remark this answers is not in the document]
Comment (alxdavids) [reply draft-irtf-cfrg-voprf.md:491]: You're right, I've just removed it.

File: draft-irtf-cfrg-voprf.md [hunk 3]
```diff
@@ -496,37 +516,30 @@ holds a secret key k for a PRF. The protocol allows the client to learn
...
+using API function calls in {{api}}.
 
 The OPRF protocol begins with V blinding its input for the OPRF
-evaluator such that it appears uniformly distributed from GG. The verifier then
-multiplies the blinded value by its secret key and returns the resulting element. 
-To finish the protocol, V then removes its blind and uses H_2 to hash the result (along 
-with a domain separating label DST) yielding an output.
-The protocol is illustrated below.
+evaluator such that it appears uniformly distributed from GG. The
+verifier then multiplies the blinded value by its secret key and returns
```
[the remark this answers is not in the document]
Comment (alxdavids) [reply draft-irtf-cfrg-voprf.md:539]: This is something that is addressed in a later PR so I will leave this for now. Essentially I just do away with all of this and just use `Client` and `Server` as `Verifier` and `Prover` doesn't even make sense in the OPRF case.

File: draft-irtf-cfrg-voprf.md [hunk 4]
```diff
@@ -535,427 +548,261 @@ The protocol is illustrated below.
...
-auxiliary input data aux.
+We restate the (V)OPRF protocol using a concrete API that we describe in
+{{algorithms}} below.
 
-### Generalized VOPRF {#general-voprf}
+Firstly, we assume that the server chooses a valid ciphersuite from
+{{ciphersuites}} and instantiates the prime-order group `GG` associated
+with the ciphersuite. It samples a PrivateKey `k` from `GF(p)`, and
+computes the public key `public_key = GG.Serialize(kG)` where
+`G=GG.generator()`. It publishes `(ciphersuite, public_key)` publicly,
```
[the remark this answers is not in the document]
Comment (alxdavids) [reply draft-irtf-cfrg-voprf.md:593]: How would you publish it non-serialized?

File: draft-irtf-cfrg-voprf.md [hunk 5]
```diff
@@ -535,427 +548,261 @@ The protocol is illustrated below.
...
-protocol above. Firstly, the server sends over an extra commitment value
-Y = kG, where G is a common generator known to both participants.
-Secondly, the server sends over both outputs from VerifiableEvaluate in
-the evaluation phase, and the client also verifies the server's output.
+Secondly, before the protocol samples an array of `ClientInput` objects
+and provides these as `ins` as their protocol input, along with a DST
+`aux`.
 
-VOPRF setup phase:
+Both participants also provide a boolean input `vv` and `vp` for the
```
[the remark this answers is not in the document]
Comment (alxdavids) [reply draft-irtf-cfrg-voprf.md:600]: I've gone through and replaced `rA` notation with `r * A` as I think this is clearer.

File: draft-irtf-cfrg-voprf.md [hunk 6]
```diff
@@ -535,427 +548,261 @@ The protocol is illustrated below.
...
- 3. Output (k,GG)
-~~~
-
 ### Blind
 
 ~~~
 Input:
 
- x: Binary string taken from {0,1}^*.
+ ClientInput inputs<1..m>
```
[the remark this answers is not in the document]
Comment (alxdavids) [reply draft-irtf-cfrg-voprf.md:703]: Ah yes, `m * sizeof(ClientInput)` is correct, I'll change. `m` is some value chosen by the client, wasn't sure how best to articulate this?

[the remark this answers is not in the document]
Comment (alxdavids) [reply draft-irtf-cfrg-voprf.md:703]: I've switched it to `Type x[m]` to indicate a fixed-size array corresponding to a size `m` chosen by the client, and added a clarifying paragraph.

File: draft-irtf-cfrg-voprf.md [hunk 7]
```diff
@@ -535,427 +548,261 @@ The protocol is illustrated below.
...
-
-## Efficiency gains with pre-processing and fixed-base blinding {#blinding}
-
-In {{oprf}} we assume that the client-side blinding is carried
-out directly on the output of H_1(x), i.e. computing rH_1(x) for some r
-<-$ GF(p). In the {{OPAQUE}} draft, it is noted that it may be more
+Let `H_1` refer to the function `GG.EncodeToGroup`, in Section
+{{algorithms}} we assume that the client-side blinding is carried out
+directly on the output of H_1(x), i.e. computing rH_1(x) for some r <-$
+GF(p). In the {{OPAQUE}} draft, it is noted that it may be more
```
[the remark this answers is not in the document]
Comment (alxdavids) [reply draft-irtf-cfrg-voprf.md:805]: Removed

File: draft-irtf-cfrg-voprf.md [hunk 8]
```diff
@@ -967,369 +814,254 @@ instantiation. We refer to the additive blinding case as fixed-base
...
-that changes is for V, the necessary computation of P does not change.
+The advantage of fixed-base blinding is that it allows the client to
+pre-process tables of blinded scalar multiplications for the fixed
+generator G. This may give it a computational efficiency advantage.
+Pre-processing also reduces the amount of computation that needs to be
+done in the online exchange. Choosing one of these values `rG` (where
+`r` is the scalar value that is used), then computing `H_1(x)+rG` is
+more efficient than computing `rH_1(x)` (one addition against log_2(r)).
+Therefore, it may be advantageous to define the OPRF and VOPRF protocols
+using additive blinding rather than multiplicative blinding. In fact,
```
[the remark this answers is not in the document]
Comment (alxdavids) [reply draft-irtf-cfrg-voprf.md:825]: Changing to additive and multiplicative

File: draft-irtf-cfrg-voprf.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:590]: As a (stupid) example, consider an app which runs the client and server in the same process. This app wouldn't need to serialize `kG` for the application to use it. Presumably the client could use whatever type `kG` is directly. 

I guess my point is that this is an application specific concern. If an application needs to publish it and send it over the wire, yeah, they'll need to serialize it. But let's not assume that's always the case.

File: draft-irtf-cfrg-voprf.md [hunk 9]
```diff
@@ -967,369 +814,254 @@ instantiation. We refer to the additive blinding case as fixed-base
...
- D: DLEQ proof (c, s).
+ PublicKey public_key
+ BlindedTokens blinded_tokens<1..m>
+ Evaluation ev
+ Scalar proof[2]
 
 Output:
 
- True if log_G(Y) == log_M(Z), False otherwise.
+ boolean verified
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:998]: Is this resolved?

File: draft-irtf-cfrg-voprf.md [hunk 10]
```diff
@@ -535,801 +546,519 @@ The protocol is illustrated below.
...
- 3.  for i in [m]:
-     1.  ri <-$ GF(p)
-     2.  Mi := rH_1(xi)
-     3.  blinds.push(ri)
-     4.  groupElems.push(Mi)
- 4.  Output (blinds, groupElems)
-~~~
-
-### VerifiableEvaluate
+The `ComputeComposites` function is a utility function used in both
```
Comment (chris-wood) [inline draft-irtf-cfrg-voprf.md:1017]:
```suggestion
`ComputeComposites` is a utility function used in both
```

**Pull request merged**